File: backend/tests/test_validation.py
"""Tests for transaction validation service."""

import pytest
from datetime import datetime, timezone, timedelta
from decimal import Decimal

from app.services.validation import (
    ValidationError,
    ValidationResult,
    TransactionValidator,
    validate_transaction,
)


class TestValidationError:
    """Tests for ValidationError dataclass."""

    def test_creation(self):
        """Test ValidationError creation."""
        error = ValidationError(field="amount", message="Amount must not be zero")
        assert error.field == "amount"
        assert error.message == "Amount must not be zero"


class TestValidationResult:
    """Tests for ValidationResult dataclass."""

    def test_empty_result_is_valid(self):
        """Test that empty result is valid."""
        result = ValidationResult()
        assert result.is_valid is True
        assert len(result.errors) == 0

    def test_add_error_makes_invalid(self):
        """Test that adding an error makes result invalid."""
        result = ValidationResult()
        result.add_error("field", "message")
        assert result.is_valid is False
        assert len(result.errors) == 1

    def test_multiple_errors(self):
        """Test multiple errors."""
        result = ValidationResult()
        result.add_error("field1", "message1")
        result.add_error("field2", "message2")
        assert result.is_valid is False
        assert len(result.errors) == 2


class TestTransactionValidatorAmount:
    """Tests for amount validation."""

    def test_valid_positive_amount(self):
        """Test valid positive amount (income)."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is True

    def test_valid_negative_amount(self):
        """Test valid negative amount (expense)."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("-50.00"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is True

    def test_zero_amount_fails(self):
        """Test that zero amount fails validation."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("0"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is False
        assert any(e.field == "amount" for e in result.errors)
        assert any("zero" in e.message.lower() for e in result.errors)

    def test_amount_below_minimum_fails(self):
        """Test that amount below minimum fails."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("0.00001"),  # Below 0.0001
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is False
        assert any(e.field == "amount" for e in result.errors)

    def test_amount_at_minimum_passes(self):
        """Test that amount at minimum passes."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("0.0001"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is True

    def test_negative_amount_at_minimum_passes(self):
        """Test that negative amount at minimum passes."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("-0.0001"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is True

    def test_amount_above_maximum_fails(self):
        """Test that amount above maximum fails."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("9999999999.9999"),  # Above max
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is False
        assert any(e.field == "amount" for e in result.errors)

    def test_amount_at_maximum_passes(self):
        """Test that amount at maximum passes."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("999999999.9999"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is True

    def test_custom_min_amount(self):
        """Test custom minimum amount."""
        validator = TransactionValidator().set_min_amount(Decimal("1.00"))
        result = validator.validate(
            amount=Decimal("0.50"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is False

    def test_custom_max_amount(self):
        """Test custom maximum amount."""
        validator = TransactionValidator().set_max_amount(Decimal("100.00"))
        result = validator.validate(
            amount=Decimal("150.00"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is False


class TestTransactionValidatorDate:
    """Tests for date validation."""

    def test_past_date_passes(self):
        """Test that past date passes."""
        validator = TransactionValidator()
        past = datetime.now(timezone.utc) - timedelta(days=30)
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=past,
        )
        assert result.is_valid is True

    def test_current_date_passes(self):
        """Test that current date passes."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is True

    def test_future_date_fails(self):
        """Test that future date fails by default."""
        validator = TransactionValidator()
        future = datetime.now(timezone.utc) + timedelta(days=1)
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=future,
        )
        assert result.is_valid is False
        assert any(e.field == "occurred_at" for e in result.errors)
        assert any("future" in e.message.lower() for e in result.errors)

    def test_future_date_allowed_when_configured(self):
        """Test that future date passes when allowed."""
        validator = TransactionValidator().set_allow_future_dates(True)
        future = datetime.now(timezone.utc) + timedelta(days=1)
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=future,
        )
        assert result.is_valid is True

    def test_naive_datetime_handled(self):
        """Test that naive datetime is handled correctly."""
        validator = TransactionValidator()
        past = datetime.now() - timedelta(days=1)  # Naive datetime
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=past,
        )
        assert result.is_valid is True


class TestTransactionValidatorIds:
    """Tests for ID validation."""

    def test_valid_wallet_id(self):
        """Test valid wallet ID."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
            wallet_id=1,
        )
        assert result.is_valid is True

    def test_zero_wallet_id_fails(self):
        """Test that zero wallet ID fails."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
            wallet_id=0,
        )
        assert result.is_valid is False
        assert any(e.field == "wallet_id" for e in result.errors)

    def test_negative_wallet_id_fails(self):
        """Test that negative wallet ID fails."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
            wallet_id=-1,
        )
        assert result.is_valid is False
        assert any(e.field == "wallet_id" for e in result.errors)

    def test_valid_category_id(self):
        """Test valid category ID."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
            category_id=1,
        )
        assert result.is_valid is True

    def test_zero_category_id_fails(self):
        """Test that zero category ID fails."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
            category_id=0,
        )
        assert result.is_valid is False
        assert any(e.field == "category_id" for e in result.errors)

    def test_negative_category_id_fails(self):
        """Test that negative category ID fails."""
        validator = TransactionValidator()
        result = validator.validate(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
            category_id=-1,
        )
        assert result.is_valid is False
        assert any(e.field == "category_id" for e in result.errors)


class TestTransactionValidatorMultiple:
    """Tests for multiple validation errors."""

    def test_multiple_errors_collected(self):
        """Test that multiple errors are collected."""
        validator = TransactionValidator()
        future = datetime.now(timezone.utc) + timedelta(days=1)
        result = validator.validate(
            amount=Decimal("0"),
            occurred_at=future,
            wallet_id=0,
            category_id=-1,
        )
        assert result.is_valid is False
        assert len(result.errors) >= 3  # amount, occurred_at, wallet_id, category_id


class TestTransactionValidatorBuilder:
    """Tests for builder pattern."""

    def test_builder_chaining(self):
        """Test that builder methods can be chained."""
        validator = (
            TransactionValidator()
            .set_min_amount(Decimal("1.00"))
            .set_max_amount(Decimal("1000.00"))
            .set_allow_future_dates(True)
        )

        # Should fail on min amount
        result = validator.validate(
            amount=Decimal("0.50"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is False

        # Should pass with valid amount
        result = validator.validate(
            amount=Decimal("500.00"),
            occurred_at=datetime.now(timezone.utc) + timedelta(days=1),  # Future allowed
        )
        assert result.is_valid is True


class TestValidateTransactionFunction:
    """Tests for convenience function."""

    def test_valid_transaction(self):
        """Test valid transaction with convenience function."""
        result = validate_transaction(
            amount=Decimal("100.00"),
            occurred_at=datetime.now(timezone.utc),
            wallet_id=1,
            category_id=1,
        )
        assert result.is_valid is True

    def test_invalid_transaction(self):
        """Test invalid transaction with convenience function."""
        result = validate_transaction(
            amount=Decimal("0"),
            occurred_at=datetime.now(timezone.utc),
        )
        assert result.is_valid is False

File: core/tests/test_validation.cpp
#include <gtest/gtest.h>
#include "validation.hpp"
#include <chrono>

using namespace cream;

// Helper to create timestamps
static Timestamp make_timestamp(int days_offset) {
    return std::chrono::system_clock::now() + std::chrono::hours(24 * days_offset);
}

// Helper to create a valid transaction
static Transaction make_valid_transaction() {
    Transaction tx;
    tx.id = 1;
    tx.wallet_id = 1;
    tx.category_id = 1;
    tx.amount = Money::from_double(50.00);
    tx.description = "Valid";
    tx.occurred_at = make_timestamp(-1);  // Yesterday
    return tx;
}

// =============================================================================
// ValidationError Tests
// =============================================================================

TEST(ValidationError, Construction) {
    ValidationError err("amount", "Amount is invalid");

    EXPECT_EQ(err.field, "amount");
    EXPECT_EQ(err.message, "Amount is invalid");
}

// =============================================================================
// ValidationResult Tests
// =============================================================================

TEST(ValidationResult, DefaultIsValid) {
    ValidationResult result;

    EXPECT_TRUE(result.is_valid());
    EXPECT_TRUE(result.valid);
    EXPECT_TRUE(result.errors.empty());
}

TEST(ValidationResult, AddErrorMakesInvalid) {
    ValidationResult result;

    result.add_error("field1", "Error 1");

    EXPECT_FALSE(result.is_valid());
    EXPECT_FALSE(result.valid);
    EXPECT_EQ(result.errors.size(), 1);
    EXPECT_EQ(result.errors[0].field, "field1");
    EXPECT_EQ(result.errors[0].message, "Error 1");
}

TEST(ValidationResult, MultipleErrors) {
    ValidationResult result;

    result.add_error("field1", "Error 1");
    result.add_error("field2", "Error 2");

    EXPECT_FALSE(result.is_valid());
    EXPECT_EQ(result.errors.size(), 2);
}

// =============================================================================
// TransactionValidator - Valid Transaction Tests
// =============================================================================

TEST(TransactionValidator, ValidTransactionPasses) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();

    ValidationResult result = validator.validate(tx);

    EXPECT_TRUE(result.is_valid());
    EXPECT_TRUE(result.errors.empty());
}

TEST(TransactionValidator, ValidNegativeAmountPasses) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.amount = Money::from_double(-50.00);  // Expense

    ValidationResult result = validator.validate(tx);

    EXPECT_TRUE(result.is_valid());
}

// =============================================================================
// TransactionValidator - Amount Validation Tests
// =============================================================================

TEST(TransactionValidator, ZeroAmountFails) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.amount = Money();  // Zero

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    EXPECT_EQ(result.errors[0].field, "amount");
}

TEST(TransactionValidator, BelowMinAmountFails) {
    TransactionValidator validator;
    validator.set_min_amount(Money::from_double(10.00));

    Transaction tx = make_valid_transaction();
    tx.amount = Money::from_double(5.00);  // Below min

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    // Should have error about being below minimum
    bool found = false;
    for (const auto& err : result.errors) {
        if (err.field == "amount" && err.message.find("below") != std::string::npos) {
            found = true;
        }
    }
    EXPECT_TRUE(found);
}

TEST(TransactionValidator, AboveMaxAmountFails) {
    TransactionValidator validator;
    validator.set_max_amount(Money::from_double(100.00));

    Transaction tx = make_valid_transaction();
    tx.amount = Money::from_double(200.00);  // Above max

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    // Should have error about exceeding maximum
    bool found = false;
    for (const auto& err : result.errors) {
        if (err.field == "amount" && err.message.find("exceeds") != std::string::npos) {
            found = true;
        }
    }
    EXPECT_TRUE(found);
}

TEST(TransactionValidator, SmallestValidAmountPasses) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.amount = Money::from_raw(1);  // 0.0001 - default minimum

    ValidationResult result = validator.validate(tx);

    EXPECT_TRUE(result.is_valid());
}

// =============================================================================
// TransactionValidator - Date Validation Tests
// =============================================================================

TEST(TransactionValidator, FutureDateFailsByDefault) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.occurred_at = make_timestamp(5);  // 5 days in future

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    EXPECT_EQ(result.errors[0].field, "occurred_at");
}

TEST(TransactionValidator, FutureDatePassesWhenAllowed) {
    TransactionValidator validator;
    validator.set_allow_future_dates(true);

    Transaction tx = make_valid_transaction();
    tx.occurred_at = make_timestamp(5);  // 5 days in future

    ValidationResult result = validator.validate(tx);

    EXPECT_TRUE(result.is_valid());
}

TEST(TransactionValidator, PastDateAlwaysPasses) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.occurred_at = make_timestamp(-365);  // 1 year ago

    ValidationResult result = validator.validate(tx);

    EXPECT_TRUE(result.is_valid());
}

// =============================================================================
// TransactionValidator - ID Validation Tests
// =============================================================================

TEST(TransactionValidator, InvalidWalletIdFails) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.wallet_id = 0;

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    EXPECT_EQ(result.errors[0].field, "wallet_id");
}

TEST(TransactionValidator, NegativeWalletIdFails) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.wallet_id = -1;

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    EXPECT_EQ(result.errors[0].field, "wallet_id");
}

TEST(TransactionValidator, InvalidCategoryIdFails) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.category_id = 0;

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    EXPECT_EQ(result.errors[0].field, "category_id");
}

TEST(TransactionValidator, NegativeCategoryIdFails) {
    TransactionValidator validator;
    Transaction tx = make_valid_transaction();
    tx.category_id = -1;

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    EXPECT_EQ(result.errors[0].field, "category_id");
}

// =============================================================================
// TransactionValidator - Builder Pattern Tests
// =============================================================================

TEST(TransactionValidator, BuilderPatternChaining) {
    TransactionValidator validator;

    // Should return reference for chaining
    auto& ref = validator
        .set_min_amount(Money::from_double(1.00))
        .set_max_amount(Money::from_double(1000.00))
        .set_allow_future_dates(true);

    EXPECT_EQ(&ref, &validator);
}

TEST(TransactionValidator, CustomMinMaxRange) {
    TransactionValidator validator;
    validator.set_min_amount(Money::from_double(10.00));
    validator.set_max_amount(Money::from_double(100.00));

    Transaction txBelow = make_valid_transaction();
    txBelow.amount = Money::from_double(5.00);

    Transaction txWithin = make_valid_transaction();
    txWithin.amount = Money::from_double(50.00);

    Transaction txAbove = make_valid_transaction();
    txAbove.amount = Money::from_double(150.00);

    EXPECT_FALSE(validator.validate(txBelow).is_valid());
    EXPECT_TRUE(validator.validate(txWithin).is_valid());
    EXPECT_FALSE(validator.validate(txAbove).is_valid());
}

// =============================================================================
// TransactionValidator - Multiple Errors Tests
// =============================================================================

TEST(TransactionValidator, CollectsMultipleErrors) {
    TransactionValidator validator;

    Transaction tx;
    tx.id = 1;
    tx.wallet_id = 0;      // Invalid
    tx.category_id = 0;    // Invalid
    tx.amount = Money();   // Zero - invalid
    tx.occurred_at = make_timestamp(5);  // Future - invalid

    ValidationResult result = validator.validate(tx);

    EXPECT_FALSE(result.is_valid());
    EXPECT_GE(result.errors.size(), 3);  // At least 3 errors
}

// =============================================================================
// Convenience Function Tests
// =============================================================================

TEST(TransactionValidator, ConvenienceFunctionWorks) {
    Transaction valid = make_valid_transaction();
    Transaction invalid = make_valid_transaction();
    invalid.amount = Money();  // Zero

    EXPECT_TRUE(validate_transaction(valid).is_valid());
    EXPECT_FALSE(validate_transaction(invalid).is_valid());
}
